clients/roscpp/src/libros/service_server_link.cpp

```diff
@@ -362,14 +368,37 @@ bool ServiceServerLink::call(const SerializedMessage& req, SerializedMessage& re
     processNextCall();
   }
 
+  bool interrupted = false;
   {
     boost::mutex::scoped_lock lock(info->finished_mutex_);
 
+    using namespace boost::chrono;
+    steady_clock::time_point now = steady_clock::now();
+
     while (!info->finished_)
     {
-      info->finished_condition_.wait(lock);
+      if (timeout > 0)
+      {
+        steady_clock::time_point wall_time = now + milliseconds(static_cast<int>(timeout * 1000));
+
+        if (info->finished_condition_.wait_until(lock, wall_time) == boost::cv_status::timeout)
```

Comment:
To make this actually work as expected you will need to include the backported boost headers and define `BOOST_THREAD_HAS_CONDATTR_SET_CLOCK_MONOTONIC` in the header file (service_server_link.h
) where the condition variable is declared. You can look at #1250 for an example.
Also it would be easier to use `wait_for` in this case...

Reply:
Done.

Thanks for the detailed comments. I wasn't aware of that issue with boost < 1.61.

Re. `wait_until` vs. `wait_for`, I think it was already changed in the original PR, but I think you're right that with `wait_for` it'd be simpler. I'd try it and update this PR.